velox/exec/PartitionedOutputBufferManager.h

```diff
@@ -40,22 +40,62 @@ struct DataAvailable {
   }
 };
 
-class DestinationBuffer {
+/// The class is used to buffer the output pages which haven't been fetched by
+/// any destination for arbitrary output.
+///
+/// NOTE: this class is not thread-safe.
+class ArbitraryBuffer {
  public:
-  void enqueue(std::shared_ptr<SerializedPage> data) {
-    // drop duplicate end markers
-    if (data == nullptr && !data_.empty() && data_.back() == nullptr) {
-      return;
-    }
+  ArbitraryBuffer() = default;
+
+  /// Returns true if this arbitrary buffer has no buffered pages.
+  bool empty() const {
+    return pages_.empty() || hasNoMoreData();
+  }
 
-    data_.push_back(std::move(data));
+  /// Returns true if this shared buffer has no more incoming pages but it can
+  /// still has buffered pages.
+  bool hasNoMoreData() const {
```

Comment:
Is null page used as and end marker?

Would it be clearer to add noMoreData() method to indicate the 'end'?

Reply:
Yes, it is. Updated.